File: actool_linux/astc_compression.py
"""ASTC-Class Compression — GPU-native quality with Apple compatibility.

ASTC (Adaptive Scalable Texture Compression) is a GPU-native texture compression
format. While true ASTC encoding requires specialized hardware/software, we can
achieve ASTC-class compression by emulating its key characteristics:

1. **Adaptive Block Size Selection**:
   - 4x4 blocks: High quality (8bpp) for edges/text
   - 6x6 blocks: Medium quality (3.56bpp)
   - 8x8 blocks: Good quality (2bpp)
   - 10x10 blocks: Lower quality (1.28bpp)
   - 12x12 blocks: Ultra compression (0.89bpp) for smooth areas

2. **Endpoint Optimization**:
   - ASTC uses color endpoints + interpolation
   - We emulate this by quantizing to block-local palettes

3. **Perceptual Weighting**:
   - ASTC optimizes for human perception
   - We use YCoCg color space + perceptual quantization

This module provides ASTC-class compression that is fully Apple-compatible
(outputs valid BGRA data that LZFSE can compress efficiently).

## New: True ASTC Integration (Experimental)

For actual ASTC encoding (requires macOS astcenc or similar), we provide
hooks to external ASTC encoders. The output is stored as opaque ASTC data
within a custom CBCK extension (codec=5, not yet supported by Apple).

For Apple compatibility, use the ASTC-class emulation mode.
"""
from __future__ import annotations

import struct
import numpy as np
from enum import IntEnum

try:
    from . import lzfse_compat as lzfse
except ImportError:
    try:
        import lzfse  # type: ignore
    except ImportError:
        lzfse = None  # type: ignore


class ASTCBlockSize(IntEnum):
    BLOCK_4x4 = 0   # 8bpp, highest quality
    BLOCK_6x6 = 1   # 3.56bpp
    BLOCK_8x8 = 2   # 2bpp, balanced
    BLOCK_10x10 = 3 # 1.28bpp
    BLOCK_12x12 = 4 # 0.89bpp, lowest quality


class ASTCClassCompressor:
    """ASTC-class compression with adaptive block sizing."""

    def __init__(self, clean_alpha: bool = True, target_bpp: float = 2.0):
        """
        Args:
            clean_alpha: Clean dirty transparency
            target_bpp: Target bits per pixel (lower = more compression)
        """
        self.clean_alpha = clean_alpha
        self.target_bpp = target_bpp

    def _select_block_size(self, block: np.ndarray) -> ASTCBlockSize:
        """Select optimal block size based on content analysis."""
        h, w = block.shape[:2]

        # Calculate edge density
        gray = block[:, :, :3].mean(axis=2).astype(np.float32)
        if w > 1 and h > 1:
            edge_x = np.abs(np.diff(gray, axis=1))
            edge_y = np.abs(np.diff(gray, axis=0))
            edge_density = (np.sum(edge_x) + np.sum(edge_y)) / (h * w * 255)
        else:
            edge_density = 0.0

        # Calculate color complexity
        flat = block.reshape(-1, 4)
        unique_ratio = len(np.unique(flat.view(np.uint32).flatten())) / len(flat)

        # Select block size based on analysis
        if edge_density > 0.15 or unique_ratio > 0.5:
            return ASTCBlockSize.BLOCK_4x4  # High detail
        elif edge_density > 0.08 or unique_ratio > 0.3:
            return ASTCBlockSize.BLOCK_6x6  # Medium detail
        elif edge_density > 0.04 or unique_ratio > 0.15:
            return ASTCBlockSize.BLOCK_8x8  # Balanced
        elif edge_density > 0.02:
            return ASTCBlockSize.BLOCK_10x10  # Low detail
        else:
            return ASTCBlockSize.BLOCK_12x12  # Smooth

    def _astc_emulate_block(self, block: np.ndarray, block_size: ASTCBlockSize) -> np.ndarray:
        """Emulate ASTC compression for a single block."""
        h, w = block.shape[:2]
        bs = block_size.value

        # ASTC uses color endpoints + interpolation
        # We emulate this with block-local palette quantization

        # Determine quantization level based on block size
        if bs == 0:  # 4x4
            levels = 32  # High quality
        elif bs == 1:  # 6x6
            levels = 16
        elif bs == 2:  # 8x8
            levels = 8
        elif bs == 3:  # 10x10
            levels = 4
        else:  # 12x12
            levels = 2  # Ultra compression

        # Quantize colors
        rgb = block[:, :, :3].astype(np.float32)
        alpha = block[:, :, 3]

        # YCoCg color space (perceptually better than RGB)
        Y = (rgb[:, :, 0] / 4.0 + rgb[:, :, 1] / 2.0 + rgb[:, :, 2] / 4.0)
        Co = (rgb[:, :, 0] / 2.0 - rgb[:, :, 2] / 2.0 + 128)
        Cg = (-rgb[:, :, 0] / 4.0 + rgb[:, :, 1] / 2.0 - rgb[:, :, 2] / 4.0 + 128)

        # Quantize (Y=chroma, Co/Cg=less important)
        scale_Y = 256 / levels
        scale_CoCg = 256 / (levels * 2)  # Chroma quantized coarser

        Y_q = np.round(Y / scale_Y) * scale_Y
        Co_q = np.round(Co / scale_CoCg) * scale_CoCg
        Cg_q = np.round(Cg / scale_CoCg) * scale_CoCg

        # RGB reconstruction
        R = Y_q + Co_q - 128
        G = Y_q + Cg_q - 128
        B = Y_q - Co_q - Cg_q + 256

        result = np.zeros_like(block)
        result[:, :, 0] = np.clip(R, 0, 255).astype(np.uint8)
        result[:, :, 1] = np.clip(G, 0, 255).astype(np.uint8)
        result[:, :, 2] = np.clip(B, 0, 255).astype(np.uint8)
        result[:, :, 3] = alpha.astype(np.uint8)

        return result

    def compress_chunk(self, chunk: np.ndarray) -> bytes:
        """Compress a chunk with ASTC-class algorithm."""
        if lzfse is None:
            return chunk.tobytes()

        h, w = chunk.shape[:2]

        # Clean alpha
        if self.clean_alpha:
            chunk = chunk.copy()
            mask = chunk[:, :, 3] == 0
            chunk[mask, :3] = 0

        # Determine optimal block size for the whole chunk
        block_size = self._select_block_size(chunk)

        # Apply ASTC-class compression
        compressed_chunk = self._astc_emulate_block(chunk, block_size)

        # LZFSE compress the result
        return lzfse.compress(compressed_chunk.tobytes())

    def compress_image(self, bgra: np.ndarray, chunk_rows: int = 256) -> bytes:
        """Compress full image."""
        if lzfse is None:
            raise RuntimeError("lzfse is required")

        h, w = bgra.shape[:2]
        chunks = []

        for y in range(0, h, chunk_rows):
            rows = min(chunk_rows, h - y)
            chunk = bgra[y:y+rows, :, :]

            compressed = self.compress_chunk(chunk)
            kcbc = b"KCBC" + struct.pack("<4I", 0, 0, rows, len(compressed)) + compressed
            chunks.append(kcbc)

        payload = b"MLEC" + struct.pack("<3I", 3, 4, len(chunks)) + b"".join(chunks)
        return payload


def astc_compress(bgra: np.ndarray, width: int, height: int, filename: str, *, scale: int = 1, target_bpp: float = 2.0) -> bytes:
    """ASTC-class compression → full CSI rendition."""
    compressor = ASTCClassCompressor(clean_alpha=True, target_bpp=target_bpp)
    bgra_arr = np.frombuffer(bgra, dtype=np.uint8).reshape(height, width, 4) if isinstance(bgra, bytes) else bgra.reshape(height, width, 4)
    payload = compressor.compress_image(bgra_arr)

    # ISTC header + TLVs
    tlvs = b"".join((
        struct.pack("<2I5I", 1001, 20, 1, 0, 0, width, height),
        struct.pack("<2I7I", 1003, 28, 1, 0, 0, 0, 0, width, height),
        struct.pack("<2I8s", 1004, 8, b"\0\0\0\0\0\0\x80?"),
        struct.pack("<2II", 1006, 4, 1),
        struct.pack("<2II", 1007, 4, width * 4),
    ))

    header = bytearray(184)
    header[:4] = b"ISTC"
    struct.pack_into("<5I", header, 4, 1, 0, width, height, scale * 100)
    header[24:28] = b"BGRA"
    struct.pack_into("<I", header, 28, 1)
    struct.pack_into("<I2H", header, 32, 0, 12, 0)
    fname_bytes = filename.encode("utf-8")[:127]
    header[40:40 + len(fname_bytes)] = fname_bytes
    struct.pack_into("<4I", header, 168, len(tlvs), 1, 0, len(payload))

    return bytes(header) + tlvs + payload

File: src/codecs/astc_compression.rs
use crate::lzfse;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ASTCBlockSize {
    Block4x4 = 0,
    Block6x6 = 1,
    Block8x8 = 2,
    Block10x10 = 3,
    Block12x12 = 4,
}

pub struct ASTCClassCompressor {
    pub clean_alpha: bool,
    pub target_bpp: f32,
}

impl Default for ASTCClassCompressor {
    fn default() -> Self {
        Self::new(true, 2.0)
    }
}

impl ASTCClassCompressor {
    pub fn new(clean_alpha: bool, target_bpp: f32) -> Self {
        Self { clean_alpha, target_bpp }
    }

    pub fn _select_block_size(&self, block_bgra: &[u8], width: usize, height: usize) -> ASTCBlockSize {
        let pixels = width * height;
        if pixels == 0 || block_bgra.len() < pixels * 4 {
            return ASTCBlockSize::Block8x8;
        }

        let (complexity, _var, _trans) = crate::astc_optimized::analyze_block_complexity(block_bgra);

        if complexity > 15.0 {
            ASTCBlockSize::Block4x4
        } else if complexity > 8.0 {
            ASTCBlockSize::Block6x6
        } else if complexity > 4.0 {
            ASTCBlockSize::Block8x8
        } else if complexity > 2.0 {
            ASTCBlockSize::Block10x10
        } else {
            ASTCBlockSize::Block12x12
        }
    }

    pub fn _astc_emulate_block(&self, block_bgra: &[u8], block_size: ASTCBlockSize) -> Vec<u8> {
        let levels = match block_size {
            ASTCBlockSize::Block4x4 => 32,
            ASTCBlockSize::Block6x6 => 16,
            ASTCBlockSize::Block8x8 => 8,
            ASTCBlockSize::Block10x10 => 4,
            ASTCBlockSize::Block12x12 => 2,
        };

        let mut out = Vec::with_capacity(block_bgra.len());
        let scale_y = 256.0 / (levels as f32);
        let scale_cocg = 256.0 / ((levels * 2) as f32);

        for px in block_bgra.chunks_exact(4) {
            let b = px[0] as f32;
            let g = px[1] as f32;
            let r = px[2] as f32;
            let a = px[3];

            let y = b / 4.0 + g / 2.0 + r / 4.0;
            let co = b / 2.0 - r / 2.0 + 128.0;
            let cg = -b / 4.0 + g / 2.0 - r / 4.0 + 128.0;

            let y_q = (y / scale_y).round() * scale_y;
            let co_q = (co / scale_cocg).round() * scale_cocg;
            let cg_q = (cg / scale_cocg).round() * scale_cocg;

            let r_rec = (y_q + co_q - 128.0).clamp(0.0, 255.0) as u8;
            let g_rec = (y_q + cg_q - 128.0).clamp(0.0, 255.0) as u8;
            let b_rec = (y_q - co_q - cg_q + 256.0).clamp(0.0, 255.0) as u8;

            out.push(b_rec);
            out.push(g_rec);
            out.push(r_rec);
            out.push(a);
        }

        out
    }

    pub fn compress_chunk(&self, bgra: &[u8], width: u32, height: u32) -> Vec<u8> {
        let mut chunk = bgra.to_vec();
        if self.clean_alpha {
            for px in chunk.chunks_exact_mut(4) {
                if px[3] == 0 {
                    px[0] = 0;
                    px[1] = 0;
                    px[2] = 0;
                }
            }
        }

        let bs = self._select_block_size(&chunk, width as usize, height as usize);
        let emulated = self._astc_emulate_block(&chunk, bs);
        lzfse::compress(&emulated)
    }

    pub fn compress_image(&self, bgra: &[u8], width: u32, height: u32) -> Vec<u8> {
        crate::cbck::encode_cbck(bgra, width, height, 4, self.clean_alpha)
    }
}

pub fn astc_compress(bgra: &[u8], width: u32, height: u32) -> Vec<u8> {
    let compressor = ASTCClassCompressor::default();
    compressor.compress_chunk(bgra, width, height)
}
